test/e2e/antreapolicy_test.go

```diff
@@ -1161,6 +1206,432 @@ func testACNPClusterGroupRefRuleIPBlocks(t *testing.T) {
+	grpName := "grp-pods-xa-then-xb"
+	grpBuilder := &GroupSpecBuilder{}
+	grpBuilder = grpBuilder.SetName(grpName).SetNamespace(namespaces["x"]).SetPodSelector(map[string]string{"pod": "a"}, nil)
+	// Update GRP Pod selector to group Pods x/b
+	updatedGrpBuilder := &GroupSpecBuilder{}
+	updatedGrpBuilder = updatedGrpBuilder.SetName(grpName).SetNamespace(namespaces["x"]).SetPodSelector(map[string]string{"pod": "b"}, nil)
+	builder := &AntreaNetworkPolicySpecBuilder{}
+	builder = builder.SetName(namespaces["x"], "anp-deny-grp-xc-to-xa-egress").
+		SetPriority(1.0).
+		SetAppliedToGroup([]ANPAppliedToSpec{{Group: grpName}})
+	builder.AddEgress(ProtocolTCP, &p80, nil, nil, nil, nil, nil, nil, nil, map[string]string{"pod": "c"}, nil,
+		nil, nil, nil, crdv1alpha1.RuleActionDrop, "", "")
+
+	reachability := NewReachability(allPods, Connected)
+	reachability.Expect(Pod(namespaces["x"]+"/a"), Pod(namespaces["x"]+"/c"), Dropped)
+
+	updatedReachability := NewReachability(allPods, Connected)
+	updatedReachability.Expect(Pod(namespaces["x"]+"/b"), Pod(namespaces["x"]+"/c"), Dropped)
+	testStep := []*TestStep{
+		{
+			"GRP Pods X/C",
+			reachability,
+			[]metav1.Object{grpBuilder.Get(), builder.Get()},
+			[]int32{80},
+			ProtocolTCP,
+			0,
+			nil,
+		},
+		{
+			"GRP Pods X/B - update",
+			updatedReachability,
+			[]metav1.Object{updatedGrpBuilder.Get()},
+			[]int32{80},
+			ProtocolTCP,
+			0,
+			nil,
+		},
+	}
+	testCase := []*TestCase{
+		{"ANP Drop Egress From Pod:x/c to Group Pod:x/a updated to Group with Pod:x/b", testStep},
+	}
+	executeTests(t, testCase)
+}
+
+func testANPGroupAppliedToPodAdd(t *testing.T, data *TestData) {
+	grpName := "grp-pod-custom-pod-xj"
+	grpBuilder := &GroupSpecBuilder{}
+	grpBuilder = grpBuilder.SetName(grpName).SetNamespace(namespaces["x"]).SetPodSelector(map[string]string{"pod": "j"}, nil)
+	builder := &AntreaNetworkPolicySpecBuilder{}
+	builder = builder.SetName(namespaces["x"], "anp-deny-grp-with-xj-to-xd-egress").
+		SetPriority(1.0).
+		SetAppliedToGroup([]ANPAppliedToSpec{{Group: grpName}})
+	builder.AddEgress(ProtocolTCP, &p80, nil, nil, nil, nil, nil, nil, nil, map[string]string{"pod": "d"}, nil,
+		nil, nil, nil, crdv1alpha1.RuleActionDrop, "", "")
+	cp := []*CustomProbe{
+		{
+			SourcePod: CustomPod{
+				Pod:    NewPod(namespaces["x"], "j"),
+				Labels: map[string]string{"pod": "j"},
+			},
+			DestPod: CustomPod{
+				Pod:    NewPod(namespaces["x"], "d"),
+				Labels: map[string]string{"pod": "d"},
+			},
+			ExpectConnectivity: Dropped,
+			Port:               p80,
+		},
+	}
+	testStep := []*TestStep{
+		{
+			"Port 80",
+			nil,
+			[]metav1.Object{grpBuilder.Get(), builder.Get()},
+			[]int32{80},
+			ProtocolTCP,
+			0,
+			cp,
+		},
+	}
+	testCase := []*TestCase{
+		{"ANP Drop Egress From Group with Pod: x/j to Pod: x/d for Pod ADD events", testStep},
+	}
+	executeTestsWithData(t, testCase, data)
+}
+
+func testANPGroupServiceRefCreateAndUpdate(t *testing.T, data *TestData) {
```

Comment:
Maybe also test the Service delete event to ensure the group and policy could be updated correctly without any nil panic.

Reply:
Also the IPBlock case?

Reply:
I just checked in `executeTests(t, testCase)` cleanup, there is fail on error for Service delete event.

Reply:
Oh I mean like you use a group to select endpoints of a Service. After this Service is deleted this group shouldn't still select those Pods.

Reply:
Done. Great suggestion, it did reveal a problem.